ntbs-service-unit-tests/DataAccess/DataQualityRepositoryTests.cs

```diff
@@ -72,6 +62,148 @@ public async Task FiltersOutNotificationsWithTheAlertTypeAlreadyPresent()
+            null, null, "John", "John", "Smith", "Smith", "2020-01-01", "2020-01-01")]
+        [InlineData(true, NotificationStatus.Notified, NotificationStatus.Notified, null, null,
+            null, null, "John", "Smith", "Smith", "John", "2020-01-01", "2020-01-01")]
+        [InlineData(true, NotificationStatus.Notified, NotificationStatus.Closed, null, null,
+            null, null, "John", "Smith", "Smith", "John", "2020-01-01", "2020-01-01")]
+        [InlineData(true, NotificationStatus.Closed, NotificationStatus.Closed, null, null,
+            null, null, "John", "Smith", "Smith", "John", "2020-01-01", "2020-01-01")]
+        [InlineData(true, NotificationStatus.Notified, NotificationStatus.Notified, "1", null,
+            null, null, "John", "Smith", "Smith", "John", "2020-01-01", "2020-01-01")]
+        [InlineData(true, NotificationStatus.Notified, NotificationStatus.Notified, "1", "2",
+            null, null, "John", "Smith", "Smith", "John", "2020-01-01", "2020-01-01")]
+
+        [InlineData(false, NotificationStatus.Deleted, NotificationStatus.Notified, null, null,
+            null, null, "John", "John", "Smith", "Smith", "2020-01-01", "2020-01-01")]
+        [InlineData(false, NotificationStatus.Notified, NotificationStatus.Denotified, null, null,
+            null, null, "John", "John", "Smith", "Smith", "2020-01-01", "2020-01-01")]
+        [InlineData(false, NotificationStatus.Notified, NotificationStatus.Draft, null, null,
+            null, null, "John", "John", "Smith", "Smith", "2020-01-01", "2020-01-01")]
+        [InlineData(false, NotificationStatus.Notified, NotificationStatus.Notified, null, null,
+            null, null, "John", "John", "Smith", "Smith", "1990-01-01", "2020-01-01")]
+        [InlineData(false, NotificationStatus.Notified, NotificationStatus.Notified, null, null,
+            null, null, "John", "John", "Smith", "Smith", null, null)]
+        [InlineData(false, NotificationStatus.Notified, NotificationStatus.Notified, null, null,
+            null, null, "John", "John", "Johnson", "Smith", "2020-01-01", "2020-01-01")]
+        [InlineData(false, NotificationStatus.Notified, NotificationStatus.Notified, null, null,
+            null, null, "John", "John", null, null, "2020-01-01", "2020-01-01")]
+        [InlineData(false, NotificationStatus.Notified, NotificationStatus.Notified, null, null,
+            null, null, "John", "James", "Smith", "Smith", "2020-01-01", "2020-01-01")]
+        [InlineData(false, NotificationStatus.Notified, NotificationStatus.Notified, null, null,
+            null, null, null, null, "Smith", "Smith", "2020-01-01", "2020-01-01")]
+        [InlineData(false, NotificationStatus.Notified, NotificationStatus.Notified, "1", "1",
+            null, null, "John", "John", "Smith", "Smith", "2020-01-01", "2020-01-01")]
+        public async Task NotificationsWithDuplicateNhsNumbers_CorrectlyFindsDuplicates(bool duplicateShouldBeFound,
```

Comment:
The fact that you have so many arguments for this method is a bit of a code smell. In this case I think it indicates that you're trying to make this test too general, and indeed it's very difficult for me to look at one of these test cases and work out what it's testing.

I think you should break this up into several tests, so that each one can require fewer parameters and have a more specific name. It's fine to leave this method as a private method that each of the tests calls out to, but just breaking up the test cases will make them a lot more readable (and therefore more maintainable).

Reply:
Yeah I agree commonisation definitely went too far here. I've split it up across the duplicate NHS numbers, and duplicate names+DoBs, and split both of these over whether we should return duplicates or not, which I think is a lot better than the `duplicateShouldBeFound` flag approach I took before. Do let me know if you can spot more/better ways to refactor this.

The only other thing I can think of is writing a test name for each of the test cases like "Does not match if they are in the same group", and adding these as messages to the assert statements, but this might be a bit overkill, I'm not sure.